third_party/move/move-vm/runtime/src/session.rs

```diff
@@ -50,57 +51,42 @@ pub struct SerializedReturnValues {
 }
 
 impl<'r, 'l> Session<'r, 'l> {
-    /// Execute a Move function with the given arguments. This is mainly designed for an external
-    /// environment to invoke system logic written in Move.
+    /// Execute a Move entry function.
     ///
-    /// NOTE: There are NO checks on the `args` except that they can deserialize into the provided
-    /// types.
-    /// The ability to deserialize `args` into arbitrary types is *very* powerful, e.g. it can
-    /// used to manufacture `signer`'s or `Coin`'s from raw bytes. It is the responsibility of the
-    /// caller (e.g. adapter) to ensure that this power is used responsibly/securely for its
-    /// use-case.
-    ///
-    /// The caller MUST ensure
-    ///   - All types and modules referred to by the type arguments exist.
-    ///   - The signature is valid for the rules of the adapter
-    ///
-    /// The Move VM MUST return an invariant violation if the caller fails to follow any of the
-    /// rules above.
-    ///
-    /// The VM will check that the function is marked as an 'entry' function.
-    ///
-    /// Currently if any other error occurs during execution, the Move VM will simply propagate that
-    /// error back to the outer environment without handling/translating it. This behavior may be
-    /// revised in the future.
-    ///
-    /// In case an invariant violation occurs, the whole Session should be considered corrupted and
-    /// one shall not proceed with effect generation.
+    /// NOTE: There are NO checks on the `args` except that they can deserialize
+    /// into the provided types. The ability to deserialize `args` into arbitrary
+    /// types is *very* powerful, e.g., it can be used to manufacture `signer`s
+    /// or `Coin`s from raw bytes. It is the responsibility of the caller to ensure
+    /// that this power is used responsibly/securely for its use-case.
     pub fn execute_entry_function(
         &mut self,
-        module: &ModuleId,
-        function_name: &IdentStr,
-        ty_args: Vec<TypeTag>,
+        func: LoadedFunction,
```

Comment:
Overall this seems like a reasonable refactoring within our current code structure. 

However I wonder if we are breaking encapsulation (even more) by not only exposing the `LoadedFunction` (supposedly a loader-internal type) to clients but also requiring them to use it. Don't have a very strong opinion on this myself, but just feel like it's worth discussing.

Reply:
Good point! I am not sure it is supposed to be an internal type? It is `pub` already but the comment says we should not rely on implementation, not exactly sure what is the intention (maybe @runtian-zhou knows).

We can encapsulate this under session API, but then we need to implement some hooks or add more functions. I do not have a strong opinion here, so let's see what everyone thinks and we can encapsulate it better.

Ideally, I think it would be nice to have
1. VM
2. Something to load a function and pass it to the VM (i.e. loader) outside.
This will make VM 2.0 integration easier and the code caching as well.

Reply:
Or we need traits/generics to properly encapsulate the APIs and not rely on concrete type

Reply:
I think it is not uncommon to have a 'handle' like LoadedFunction exposed via the public APIs. Its definitely cleaner then the break down in low-level components like ModuleId, IdentStr, etc.